### PR TITLE
add skeleton of action analyzer

### PR DESCRIPTION
**Pull request by ycheng35xo**

### CODE REVIEW

File: assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  spark.executor.cores: 4
+  spark.executor.memory: 28g
+  spark.executor.instances: 1
+  spark.dynamicAllocation.enabled: True
+  spark.dynamicAllocation.minExecutors: 1
+  spark.dynamicAllocation.maxExecutors: 1 # avoid AOAI API limits by running only one executor at a time
+  spark.jars.packages: com.microsoft.azure:synapseml_2.12:0.11.0-26-52919ce4-SNAPSHOT,org.apache.spark:spark-avro_2.12:3.3.1
+  spark.jars.repositories: https://mmlspark.azureedge.net/maven
+  spark.jars.excludes: org.scala-lang:scala-reflect,org.apache.spark:spark-tags_2.12,org.scalactic:scalactic_2.12,org.scalatest:scalatest_2.12,com.fasterxml.jackson.core:jackson-databind
+  spark.yarn.user.classpath.first: true
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml:30]: do we need this?

File: assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+conf:
+  spark.driver.cores: 4
+  spark.driver.memory: 28g
+  spark.executor.cores: 4
+  spark.executor.memory: 28g
+  spark.executor.instances: 1
+  spark.dynamicAllocation.enabled: True
+  spark.dynamicAllocation.minExecutors: 1
+  spark.dynamicAllocation.maxExecutors: 1 # avoid AOAI API limits by running only one executor at a time
+  spark.jars.packages: com.microsoft.azure:synapseml_2.12:0.11.0-26-52919ce4-SNAPSHOT,org.apache.spark:spark-avro_2.12:3.3.1
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml:27]: do we need this?

File: assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  spark.driver.userClassPathFirst: false
+  spark.executor.userClassPathFirst: false
+  spark.synapse.library.python.env: |
+    channels:
+      - defaults
+      - conda-forge
+    dependencies:
+      - python=3.10
+      - pip:
+          - azure-cli-core~=2.45.0
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml:42]: could you refer shreeya's PR?

File: assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml (hunk 2 above)
Comment (yunjie-hub) [reply assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml:27]: Similar to other components

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_correlation_test/run.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    actions_row = []
+    for index in index_set:
+        for group in group_set:
+            print("Group: ", group)
+            metrics = group.split("_")[0]
+            good_group_name = f"{metrics}_good_group"
+            if group == good_group_name:
+                print(f"Skip {good_group_name}, only do t-test for bad group")
+                continue
+            index_df = pdf[pdf['index_id'] == index]
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/action_analyzer/action_analyzer_correlation_test/run.py:52]: index_id, index_score,group_list, those hard coded const value, make sure we don't have those columns

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    _APITokenManager,
+    _WorkspaceConnectionTokenManager,
+    _HTTPClientWithRetry,
+    _check_and_format_azure_endpoint_url,
+    _request_api,
+    get_openai_request_args
+)
+
+
+Retrieval_document_RELEVANCE_TEMPLATE = "\n\n".join(
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py:41]: should we define it in a file

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+            # print("rating=", rating, type(rating))
+            ratings.append(rating)
+        except Exception as e:  # noqa: B902
+            response["finish_reason"] = ["error"]
+            response["error"] = [str(e)]
+            raise e
+    return ratings
+
+
+def _query_relevance_score(
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py:202]: why do we need to have a _query_relevance_score to wrap _query_relevance_scores

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+            response["response_time_sec"] = time_taken
+            print(response["samples"][0])
+            rating = json.loads(response["samples"][0])["rating"]
+            # print("===response===")
+            # print("rating=", rating, type(rating))
+            ratings.append(rating)
+        except Exception as e:  # noqa: B902
+            response["finish_reason"] = ["error"]
+            response["error"] = [str(e)]
+            raise e
```
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py:198]: What kind of exception do we have?

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py (hunk 6 above)
Comment (yunjie-hub) [reply assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py:202]: This code will deleted, right?

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py (hunk 7 above)
Comment (yunjie-hub) [inline assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py:198]: need to consider what will be the exception, and should we fail the job or let the job continue

File: assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml (hunk 2 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/model_monitor_action_analyzer/action_analyzer_correlation_test/spec.yaml]: I removed all of them

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py (hunk 6 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/action_analyzer/action_analyzer_metrics_calculation/run.py]: removed.

File: assets/model_monitoring/components/src/action_analyzer/action_analyzer_correlation_test/run.py (hunk 4 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/action_analyzer/action_analyzer_correlation_test/run.py]: Make them as constants. Also in the first step, only taking the columns from GSQ, which will not have extra fields like these

**Pull request merged**